### PR TITLE
Fix `SimplifyArmIdentity` MIR opt

### LINKED ISSUE #72800: `SimplifyArmIdentity` can cause use-after-moves in MIR
Reduced example diff from `<core::slice::Iter<'a, T> as Iterator>::find_map`:

```diff
-// MIR for `slice::<impl at src/libcore/slice/mod.rs:3293:9: 3475:10>::find_map` before SimplifyArmIdentity
+// MIR for `slice::<impl at src/libcore/slice/mod.rs:3293:9: 3475:10>::find_map` after SimplifyArmIdentity
 
 fn slice::<impl at src/libcore/slice/mod.rs:3293:9: 3475:10>::find_map(_1: &mut slice::Iter<T>, _2: F) -> option::Option<B> {
     debug self => _1;                    // in scope 0 at src/libcore/slice/mod.rs:3420:31: 3420:40
     debug f => _2;                       // in scope 0 at src/libcore/slice/mod.rs:3420:42: 3420:47
     let mut _0: option::Option<B>;       // return place in scope 0 at src/libcore/slice/mod.rs:3420:55: 3420:64
     let _3: ();                          // in scope 0 at src/libcore/slice/mod.rs:3425:17: 3429:18
     let mut _4: ();                      // in scope 0 at src/libcore/slice/mod.rs:3420:13: 3431:14
     let mut _5: option::Option<&T>;      // in scope 0 at src/libcore/slice/mod.rs:3425:37: 3425:48
     let mut _6: &mut slice::Iter<T>;     // in scope 0 at src/libcore/slice/mod.rs:3425:37: 3425:41
     let mut _7: isize;                   // in scope 0 at src/libcore/slice/mod.rs:3425:27: 3425:34
     let _8: &T;                          // in scope 0 at src/libcore/slice/mod.rs:3425:32: 3425:33
     let mut _9: option::Option<B>;       // in scope 0 at src/libcore/slice/mod.rs:3426:38: 3426:42
     let mut _10: &mut F;                 // in scope 0 at src/libcore/slice/mod.rs:3426:38: 3426:39
     let mut _11: (&T,);                  // in scope 0 at src/libcore/slice/mod.rs:3426:38: 3426:42
     let mut _12: &T;                     // in scope 0 at src/libcore/slice/mod.rs:3426:40: 3426:41
     let mut _13: isize;                  // in scope 0 at src/libcore/slice/mod.rs:3426:28: 3426:35
     let mut _15: !;                      // in scope 0 at src/libcore/slice/mod.rs:3426:43: 3428:22
     let mut _16: B;                      // in scope 0 at src/libcore/slice/mod.rs:3427:37: 3427:38
     let mut _17: !;                      // in scope 0 at src/libcore/slice/mod.rs:3425:17: 3429:18
     let mut _18: isize;                  // in scope 0 at src/libcore/slice/mod.rs:3429:17: 3429:18
     let mut _19: isize;                  // in scope 0 at src/libcore/slice/mod.rs:3429:17: 3429:18
     scope 1 {
         debug x => _8;                   // in scope 1 at src/libcore/slice/mod.rs:3425:32: 3425:33
         let _14: B;                      // in scope 1 at src/libcore/slice/mod.rs:3426:33: 3426:34
         scope 2 {
             debug y => _14;              // in scope 2 at src/libcore/slice/mod.rs:3426:33: 3426:34
         }
     }
 
     bb0: {
         StorageLive(_3);                 // scope 0 at src/libcore/slice/mod.rs:3425:17: 3429:18
         goto -> bb1;                     // scope 0 at src/libcore/slice/mod.rs:3425:17: 3429:18
     }

     // <...>
 
     bb7: {
-        StorageLive(_14);                // scope 1 at src/libcore/slice/mod.rs:3426:33: 3426:34
-        _14 = move ((_9 as Some).0: B);  // scope 1 at src/libcore/slice/mod.rs:3426:33: 3426:34
-        StorageLive(_16);                // scope 2 at src/libcore/slice/mod.rs:3427:37: 3427:38
-        _16 = move _14;                  // scope 2 at src/libcore/slice/mod.rs:3427:37: 3427:38
-        ((_0 as Some).0: B) = move _16;  // scope 2 at src/libcore/slice/mod.rs:3427:32: 3427:39
-        discriminant(_0) = 1;            // scope 2 at src/libcore/slice/mod.rs:3427:32: 3427:39
-        StorageDead(_16);                // scope 2 at src/libcore/slice/mod.rs:3427:38: 3427:39
-        StorageDead(_14);                // scope 1 at src/libcore/slice/mod.rs:3428:21: 3428:22
+        _0 = move _9;                    // scope 2 at src/libcore/slice/mod.rs:3427:32: 3427:39
         _18 = discriminant(_9);          // scope 1 at src/libcore/slice/mod.rs:3429:17: 3429:18
         StorageDead(_9);                 // scope 1 at src/libcore/slice/mod.rs:3429:17: 3429:18
         StorageDead(_8);                 // scope 0 at src/libcore/slice/mod.rs:3429:17: 3429:18
         StorageDead(_5);                 // scope 0 at src/libcore/slice/mod.rs:3429:17: 3429:18
         StorageDead(_3);                 // scope 0 at src/libcore/slice/mod.rs:3429:17: 3429:18
         goto -> bb9;                     // scope 0 at src/libcore/slice/mod.rs:3431:13: 3431:14
     }
 
     bb9: {
         return;                          // scope 0 at src/libcore/slice/mod.rs:3431:14: 3431:14
     }
 }
```

Here, the `move _9` makes `_9` uninitialized (in `MaybeInitializedLocals`), but the next statement immediately uses it again via `discriminant(_9)`.

I do not believe that this is currently a soundness issue, as no analyses other than the generator transform (which runs earlier) should rely on this data. It *is* a problem for https://github.com/rust-lang/rust/pull/72632 though, since that runs after `SimplifyArmIdentity`. Granted, the example above is unlikely to be problematic even for that, but other cases might be.

### LINKED ISSUE #77359: Segfaults/corruption when reading an enum in release mode
Reproduction here: https://github.com/cormacrelf/minimal-sigsegv-rust

There's more info in the readme on that repo. Essentially, I have some code that segfaults while debug-printing an enum. The enum has one `Output(String)` variant, followed by 11 empty variants. The repro only creates one of the empty ones, but it appears the debug routine attempts to read the string variant. It occurs in the context of some mutually recursive functions building a fairly simple datastructure by walking another. I have tried hard to minimise the repro but it could be better, but it's a bit fragile so I don't know what else I can remove. With a quick glance at the commit that bisect-rustc found, it looks like it could be related to optimising match arms, so that could be a start.

Edit: for some linkage, the commit below merges #76308, which 're-enables SimplifyArmIdentity'.

### cargo bisect-rustc

searched nightlies: from nightly-2020-05-08 to nightly-2020-09-23
regressed nightly: nightly-2020-09-09
searched commits: from https://github.com/rust-lang/rust/commit/0e2c1281e909ca38479b97962fc9248f75d66412 to https://github.com/rust-lang/rust/commit/5099914a16a215794ad243df0cc7a05d91d168e0
regressed commit: https://github.com/rust-lang/rust/commit/5a6b426e3471382c0385c11b09c2d6b37f70ac49

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.5.2</summary>

Host triple: x86_64-apple-darwin
Reproduce with:
```bash
cargo bisect-rustc --start 2020-05-08 --end 2020-09-23 --with-cargo --access github -- run --release
```
</details>

Note that the `ref_sequence` function is the one in which the debug segfaults, and it is debugging a larger structure. I have also seen it segfault when cloning just the `EdgeData` enum, which is the String+11x one mentioned above.

<details><summary>Backtrace from rust-lldb</summary>
<p>

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x0000000100002588 minimal`_$LT$$RF$T$u20$as$u20$core..fmt..Debug$GT$::fmt::h797ace08a5294f9f + 24
    frame #1: 0x000000010002e163 minimal`core::fmt::builders::DebugTuple::field::h5368cface0fdb245 [inlined] core::fmt::builders::DebugTuple::field::_$u7b$$u7b$closure$u7d$$u7d$::hb6c824bec58214a9 at builders.rs:347:17 [opt]
    frame #2: 0x000000010002e115 minimal`core::fmt::builders::DebugTuple::field::h5368cface0fdb245 [inlined] core::result::Result$LT$T$C$E$GT$::and_then::h2b3c0ffee0d30aef at result.rs:708 [opt]
    frame #3: 0x000000010002e109 minimal`core::fmt::builders::DebugTuple::field::h5368cface0fdb245 at builders.rs:334 [opt]
    frame #4: 0x000000010000459f minimal`_$LT$$RF$T$u20$as$u20$core..fmt..Debug$GT$::fmt::h2b06d94161eb79e0 + 95
    frame #5: 0x000000010002e163 minimal`core::fmt::builders::DebugTuple::field::h5368cface0fdb245 [inlined] core::fmt::builders::DebugTuple::field::_$u7b$$u7b$closure$u7d$$u7d$::hb6c824bec58214a9 at builders.rs:347:17 [opt]
    frame #6: 0x000000010002e115 minimal`core::fmt::builders::DebugTuple::field::h5368cface0fdb245 [inlined] core::result::Result$LT$T$C$E$GT$::and_then::h2b3c0ffee0d30aef at result.rs:708 [opt]
    frame #7: 0x000000010002e109 minimal`core::fmt::builders::DebugTuple::field::h5368cface0fdb245 at builders.rs:334 [opt]
    frame #8: 0x00000001000037e8 minimal`_$LT$minimal..ref_ir..RefIR$u20$as$u20$core..fmt..Debug$GT$::fmt::hfde74232d0e0d5d0 + 104
    frame #9: 0x000000010002e910 minimal`core::fmt::write::h0ce880d33cd2a300 at mod.rs:1080:17 [opt]
    frame #10: 0x0000000100015034 minimal`_$LT$$RF$std..io..stdio..Stderr$u20$as$u20$std..io..Write$GT$::write_fmt::h2508a17060d785a3 [inlined] std::io::Write::write_fmt::hdeaf6be33655ecfd at mod.rs:1517:15 [opt]
    frame #11: 0x0000000100014fe9 minimal`_$LT$$RF$std..io..stdio..Stderr$u20$as$u20$std..io..Write$GT$::write_fmt::h2508a17060d785a3 at stdio.rs:838 [opt]
    frame #12: 0x0000000100015596 minimal`std::io::stdio::_eprint::h8f478498c62451a2 [inlined] _$LT$std..io..stdio..Stderr$u20$as$u20$std..io..Write$GT$::write_fmt::h4bdcbad34c0de397 at stdio.rs:812:9 [opt]
    frame #13: 0x0000000100015539 minimal`std::io::stdio::_eprint::h8f478498c62451a2 [inlined] std::io::stdio::print_to::_$u7b$$u7b$closure$u7d$$u7d$::h805d34d32e3d72cc at stdio.rs:947 [opt]
    frame #14: 0x0000000100015404 minimal`std::io::stdio::_eprint::h8f478498c62451a2 [inlined] std::thread::local::LocalKey$LT$T$GT$::try_with::h81d8f4dd4c7ef03e at local.rs:271 [opt]
    frame #15: 0x00000001000153dc minimal`std::io::stdio::_eprint::h8f478498c62451a2 [inlined] std::io::stdio::print_to::h3e10e6f958040f4c at stdio.rs:936 [opt]
    frame #16: 0x00000001000153dc minimal`std::io::stdio::_eprint::h8f478498c62451a2 at stdio.rs:975 [opt]
    frame #17: 0x0000000100002c3b minimal`minimal::disamb::element_ref_ir_impl::h81621c2e3d855e9c + 1035
    frame #18: 0x000000010000423f minimal`minimal::main::h79d5d72421065e13 + 1375
    frame #19: 0x00000001000022fa minimal`std::sys_common::backtrace::__rust_begin_short_backtrace::h1ea652726f9417f0 + 10
    frame #20: 0x000000010000235c minimal`std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h254a4aa9b725da66 (.llvm.9306634996709519214) + 12
    frame #21: 0x000000010001a290 minimal`std::rt::lang_start_internal::h5e1feb19f4099625 [inlined] core::ops::function::impls::_$LT$impl$u20$core..ops..function..FnOnce$LT$A$GT$$u20$for$u20$$RF$F$GT$::call_once::h31ef09f9bee131c1 at function.rs:259:13 [opt]
    frame #22: 0x000000010001a286 minimal`std::rt::lang_start_internal::h5e1feb19f4099625 [inlined] std::panicking::try::do_call::h693e4d9f7366f6d8 at panicking.rs:381 [opt]
    frame #23: 0x000000010001a286 minimal`std::rt::lang_start_internal::h5e1feb19f4099625 [inlined] std::panicking::try::h58e51096fb939dc8 at panicking.rs:345 [opt]
    frame #24: 0x000000010001a286 minimal`std::rt::lang_start_internal::h5e1feb19f4099625 [inlined] std::panic::catch_unwind::h4406f92fae32aea8 at panic.rs:382 [opt]
    frame #25: 0x000000010001a286 minimal`std::rt::lang_start_internal::h5e1feb19f4099625 at rt.rs:51 [opt]
    frame #26: 0x0000000100004539 minimal`main + 41
    frame #27: 0x00007fff7f3603d5 libdyld.dylib`start + 1
    frame #28: 0x00007fff7f3603d5 libdyld.dylib`start + 1
```

</p>
</details>

### LINKED ISSUE #78628: ICE: broken mir with -Zunsound-mir-opts:  simplify arm identity
### Code
from `./src/test/ui/mir/issue-77359-simplify-arm-identity.rs`
```Rust
// run-pass

#![allow(dead_code)]

#[derive(Debug)]
enum MyEnum {
    Variant2,
    Variant3,
}

fn f(arg3: bool) -> MyStruct {
	match if arg3 { Some(MyEnum::Variant3) } else { None } {
		Some(t) => {
			let ah = t;
			return MyStruct(Some(ah));
		}
		_ => MyStruct(None)
	}
}

#[derive(Debug)]
struct MyStruct(Option<MyEnum>);

fn main() {
    f(true);
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.49.0-nightly (1d5b7c3c9 2020-11-01)
binary: rustc
commit-hash: 1d5b7c3c96a86c96f1fbf1441475028df94d165a
commit-date: 2020-11-01
host: x86_64-unknown-linux-gnu
release: 1.49.0-nightly
```

### Error output

```
error: internal compiler error: broken MIR in Item(WithOptConstParam { did: DefId(0:3 ~ issue_77359_simplify_arm_identity[317d]::f), const_param_did: None }) (end of phase Optimization) at bb5[0]:
use of local _5, which has no storage here
  --> ./src/test/ui/mir/issue-77359-simplify-arm-identity.rs:15:20
   |
15 |             return MyStruct(Some(ah));
   |                             ^^^^^^^^
   |
   = note: delayed at compiler/rustc_mir/src/transform/validate.rs:171:36

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:958:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.49.0-nightly (1d5b7c3c9 2020-11-01) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z unsound-mir-opts

query stack during panic:
end of query stack
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: broken MIR in Item(WithOptConstParam { did: DefId(0:3 ~ issue_77359_simplify_arm_identity[317d]::f), const_param_did: None }) (end of phase Optimization) at bb5[0]:
use of local _5, which has no storage here
  --> ./src/test/ui/mir/issue-77359-simplify-arm-identity.rs:15:20
   |
15 |             return MyStruct(Some(ah));
   |                             ^^^^^^^^
   |
   = note: delayed at compiler/rustc_mir/src/transform/validate.rs:171:36

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:958:13
stack backtrace:
   0:     0x7ffaf14ba5f0 - std::backtrace_rs::backtrace::libunwind::trace::h448c56741b6011d3
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/../../backtrace/src/backtrace/libunwind.rs:100:5
   1:     0x7ffaf14ba5f0 - std::backtrace_rs::backtrace::trace_unsynchronized::h2d26397c5720fdbb
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7ffaf14ba5f0 - std::sys_common::backtrace::_print_fmt::h720a2f61f75b9f58
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7ffaf14ba5f0 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h8d62bbfda6d5c836
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x7ffaf152a52c - core::fmt::write::h1857a60b204f1b6a
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/core/src/fmt/mod.rs:1078:17
   5:     0x7ffaf14ac252 - std::io::Write::write_fmt::h16a9fe7680ac245c
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/io/mod.rs:1518:15
   6:     0x7ffaf14bf5a5 - std::sys_common::backtrace::_print::h9ba51a6db618de7d
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x7ffaf14bf5a5 - std::sys_common::backtrace::print::h5f20e41c85e91716
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x7ffaf14bf5a5 - std::panicking::default_hook::{{closure}}::h5b7294ca19e8edab
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/panicking.rs:208:50
   9:     0x7ffaf14bf248 - std::panicking::default_hook::hb3948d1f74b6ff4c
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/panicking.rs:227:9
  10:     0x7ffaf1d8e158 - rustc_driver::report_ice::h1ddbc5609a6e1f37
  11:     0x7ffaf14bfea6 - std::panicking::rust_panic_with_hook::h55d23fef0ad751bc
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/panicking.rs:597:17
  12:     0x7ffaf14bf9c7 - std::panicking::begin_panic_handler::{{closure}}::h56ae74c6dea6e141
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/panicking.rs:499:13
  13:     0x7ffaf14baa8c - std::sys_common::backtrace::__rust_end_short_backtrace::h3c4721f56689fae4
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/sys_common/backtrace.rs:141:18
  14:     0x7ffaf14bf929 - rust_begin_unwind
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/panicking.rs:495:5
  15:     0x7ffaf14bf8db - std::panicking::begin_panic_fmt::h447a9fe10c3ac155
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/panicking.rs:437:5
  16:     0x7ffaf4dff1c1 - rustc_errors::HandlerInner::flush_delayed::hcdae04493e012a6e
  17:     0x7ffaf4dfbb91 - <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop::h27a485c12bfd5a8a
  18:     0x7ffaf1d9b606 - core::ptr::drop_in_place::hac2c86d61f6a657d
  19:     0x7ffaf1da0796 - <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop::h60c965ef9381f1d3
  20:     0x7ffaf1d46b2d - core::ptr::drop_in_place::h11c56312c481deac
  21:     0x7ffaf1d36a2c - rustc_span::with_source_map::h4e49cbd9a39d01b4
  22:     0x7ffaf1da70c1 - scoped_tls::ScopedKey<T>::set::hca96dcfe1d32eebf
  23:     0x7ffaf1daed75 - std::sys_common::backtrace::__rust_begin_short_backtrace::h21cfd4855c5058aa
  24:     0x7ffaf1d17188 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h2df6ccf8e166462a
  25:     0x7ffaf14ceeaa - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h9e7afb7a0a438236
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/alloc/src/boxed.rs:1307:9
  26:     0x7ffaf14ceeaa - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h70c646c4271337a1
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/alloc/src/boxed.rs:1307:9
  27:     0x7ffaf14ceeaa - std::sys::unix::thread::Thread::new::thread_start::h35d2b8d36f210d02
                               at /rustc/1d5b7c3c96a86c96f1fbf1441475028df94d165a/library/std/src/sys/unix/thread.rs:89:17
  28:     0x7ffaf13ca3e9 - start_thread
  29:     0x7ffaf12e7293 - __GI___clone
  30:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.49.0-nightly (1d5b7c3c9 2020-11-01) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z unsound-mir-opts

query stack during panic:
end of query stack
```

</p>
</details>

### PR DESCRIPTION
**Pull request by JakobDegen**
Closes #77359
Closes #72800
Closes #78628

This is a near full re-write of the pass. The StorageLive issue would have been relatively easy to fix, but the use after move issue was more complicated. The new version of the pass documents very explicitly how `copy`/`move` things are handled, and takes care to uphold that. This version also increases the flexibility of the pass. We allow any number of fields (instead of just one) and also allow moving into ourselves.

This pass does a little less than the original version. In particular, it leaves all the storage and some of the temporary assignments in. I do think that it is in general the right idea to not try and also remove those here. I've written #94118 which is supposed to run soon after this and is designed to remove these kinds of things. Unfortunately, this also means that a lot of the tests around `SimplifyBranchSame` break, because that optimization no longer runs. I've gone and deleted these tests for now. I don't know if we even want to keep the `SimplifyBranchSame` pass at all though; this pass is enough to get LLVM to output the right thing, so it might be enough to let the remaining issues get swept up in DSE + SimplifyCFG or something like that.

The diff is pretty big, but that's mostly tests. I've organized the commits so that hopefully the changes to the test suite are pretty easy to follow.

cc @wesleywiser who seems to have been the last person to touch this

r? rust-lang/mir-opt

### CODE REVIEW

File: compiler/rustc_mir_transform/src/simplify_try.rs [hunk 1]
```diff
@@ -11,529 +11,435 @@
...
+/// ```ignore (syntax-highlighting-only)
+/// TMP_1 = move? ((SRC as Variant).FIELD);
+/// TMP_2 = move? TMP_1;
+/// ((DEST as Variant).FIELD) = move? TMP_2;
 /// ```
 ///
-/// into:
+/// where `Variant`, `SRC`, and `DEST` must be consistent across all bundles, while `TMP_1`,
+/// `TMP_2`, and `FIELD` must vary across all bundles. The actual number of `TMP_*`s that we require
+/// is not fixed; we allow arbitrary many, including possibly zero.
```
Comment (JakobDegen) [inline compiler/rustc_mir_transform/src/simplify_try.rs:36]: Just saw the type in "arbitrarily," will fix whenever I push updates next

**Comment by JakobDegen** [comment]
*Sigh*, this will have to wait until tomorrow

**Comment by JakobDegen** [comment]
Was missing a single `?`. (well, was missing a `#[must_use]` that allowed me to miss a `?`)

**Comment by JakobDegen** [comment]
Oop, didn't know there was a codegen test. I've modified it for now by turning on LLVM optimizations. With `SimplifyBranchSame` no longer running, its expected that there's still a branch in the unoptimized LLVM-IR, but LLVM does remove it.

**Comment by JakobDegen** [comment]
Hm, that test doesn't seem to run locally by default. Is there anyway to get `./x.py test src/test/mir-opt` to run this? Even if it doesn't run the resulting executable, it should at least be able to generate the `diff` files

File: compiler/rustc_mir_transform/src/simplify_try.rs [hunk 2]
```diff
@@ -11,529 +11,436 @@
...
+/// assume that anything `copy`ed may be used again later. Making sure we don't miscompile something
+/// here requires care. We impose the following rules:
+///
+/// 1. All assignments out of the source must have the same move mode. This is the move mode that is
+///    used for the single assignment that is found in the new output statement. If that move mode
+///    then turns out to be `copy`, we additionally require the type of `src` to satisfy a `Copy`
+///    bound.
+/// 2. For each bundle, if any assignment to a temporary is via `copy`, then this is sufficient to
+///    prove that the field is `Copy`. In this case, we keep all assignments to temporaries
+///    associated with this bundle. The move mode of the assignment out of `src` is forced to be
```
[the remark this answers is not in the document]
Comment (JakobDegen) [reply compiler/rustc_mir_transform/src/simplify_try.rs:74]: @cjgillot see #94118 , which would run after this, and is a smaller (and less perf intensive) hammer for the common case

**Comment by cjgillot** [comment]
You need to specify another `--target i686-unknown-linux-gnu` (or another well-chosen target) to `x.py` for it to run 32bit mir-opts.

File: compiler/rustc_mir_transform/src/simplify_try.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/simplify_try.rs:74]: Could you demonstrate what mir-opt tests look like with both optimizations running?
Other option: we are creating a list of `StorageDead` statements to be moved. Can it be used to kill off unused `Copy` locals?

[the remark this answers is not in the document]
Comment (JakobDegen) [reply compiler/rustc_mir_transform/src/simplify_try.rs:74]: Actually, I just had a different idea. If that doesn't pan out though, I'll add a test demonstrating this

**Comment by JakobDegen** [comment]
@cjgillot I've added a call to `simplify_locals` at the end of the opt if it fired. This unfortunately degrades the quality of the tests, but I don't really see a better solution.

**Comment by cjgillot** [comment]
> @cjgillot I've added a call to `simplify_locals` at the end of the opt if it fired. This unfortunately degrades the quality of the tests, but I don't really see a better solution.

Thanks.
Indeed, `simplify_locals` cleans up cruft left by several passes, which makes the tests diffs confusing.
What about the idea you talked about in https://github.com/rust-lang/rust/pull/94177#discussion_r811373952?
What about using the `StorageDead` information?
Also, could you reinstate the `SimplifyBranchSame` tests in place?

**Comment by JakobDegen** [comment]
> 

My hope is that I can also start working on dest prop and fix that so that it can run immediately after. At that point, I can remove the `simplify_locals` call because dest prop should pick everything up on its own. This would improve the tests and avoid issues with overhead. I can look at using information from the accumulated `StorageDead` as well, but I don't know if that will be worth it if dest prop is fixed.

**Comment by JakobDegen** [comment]
I've added the tests back in. Would be interested in seeing the results of a perf run of this, I don't expect it to be too expensive, but I have no idea how often it will fire.

**Comment by JakobDegen** [comment]
I found a comment in other code indicating that self-assignment is UB. I can push an update to this pass that avoids emitting such a self-assignment, but is that the case? If so, why?

**Comment by JakobDegen** [comment]
I've converted this to a draft to keep it from being merged, pending #94590 . That will require some adjustments to this pass. Additionally there are two additional checks that need to be introduced for this to be correct: The type needs to be an ADT, not a generator, and the move into self needs to be avoided.

**Comment by JakobDegen** [comment]
I have a version of this that I think is ready to review, it's only blocked on #96090 . Once that lands, I'll push updates

**Comment by Mark-Simulacrum** [comment]
@JakobDegen FYI: It looks like https://github.com/rust-lang/rust/pull/96090 has landed at this point.

**Comment by JakobDegen** [comment]
Yeah, I'm still waiting for more changes though. I can close this if it's better to not have it up

**Comment by JakobDegen** [comment]
I'm closing this PR because it is still unsound in some very hard to fix ways.

**Pull request closed without merging**